skimage/future/trainable_segmentation.py

```diff
@@ -0,0 +1,164 @@
+from skimage.feature import multiscale_basic_features
+
+try:
+    from sklearn.exceptions import NotFittedError
+    from sklearn.ensemble import RandomForestClassifier
+    has_sklearn = True
+except ImportError:
+    has_sklearn = False
+
+    class NotFittedError(Exception):
+        pass
+
+
+class TrainableSegmenter(object):
+    """
+    Estimator for classifying pixels.
+
+    Parameters
+    ----------
+    clf : classifier object, optional
+        classifier object, exposing a ``fit`` and a ``predict`` method as in
+        scikit-learn's API, for example an instance of
+        ``RandomForestClassifier`` or ``LogisticRegression`` classifier.
+    features_func : function, optional
+        function computing features on all pixels of the image, to be passed
+        to the classifier. The output should be of shape
+        ``(m_features, *labels.shape)``. If None,
+        :func:`skimage.segmentation.multiscale_basic_features` is used.
+
+    Methods
+    -------
+    compute_features
+    fit
+    predict
+    """
+
+    def __init__(self, clf=None, features_func=None):
+        if clf is None:
+            if has_sklearn:
+                self.clf = RandomForestClassifier(n_estimators=100, n_jobs=-1)
+            else:
+                raise ImportError(
+                    "Please install scikit-learn or pass a classifier instance"
+                    "to TrainableSegmenter."
+                )
+        else:
+            self.clf = clf
+        self.features_func = features_func
+
+    def compute_features(self, image):
+        if self.features_func is None:
+            self.features_func = multiscale_basic_features
```

Comment:
```suggestion
```
This part has to be in the `__init__`

Reply:
Thanks for the review @rfezzani; having `features` a `property` is an interesting idea but I'm not sure how this would work out. The syntax I can think of would be like:
```python
@property
    def features(self):
        return self.features_func(self.image)

    @features.setter
    def features(self, image):
        return self.features_func(image)
```
but then in the `fit` method I would need to write (using the setter):
```python
self.features = image
```
which looks misleading to me. Or did you have something else in mind? Thanks!

Reply:
You need to store `image` in `self` as you did in the definition of the `features` property :wink::
```python
@property
def features(self):
    return self.features_func(self.image)

def fit(self, labels, image=None):
    self.image = image
    clf = fit_segmenter(labels, self.features, self.clf)
```

but the drawback is that you end up calling `self.features_func` whenever you access `self.features` :slightly_frowning_face: .
Using `property` is probably not a good idea :slightly_smiling_face: